### ISSUE TITLE
test_commitlog_shutdown_during_wait: log shutdown timed out.

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#10805]
Stumbled upon in https://jenkins.scylladb.com/job/releng/job/Scylla-CI/827/artifact/testlog/x86_64_release/commitlog_test.test_commitlog_shutdown_during_wait.643.log
Scylla version 771dabb5782713d4de84102dc3272a50c0a7df52 (#10657) based on c0a09669c156b46e30b1d644d512160417778d27
```
random-seed=2322839544
test/boost/commitlog_test.cc(880): fatal error: in "test_commitlog_shutdown_during_wait": log shutdown timed out. maybe it is deadlocked... Will not free log. ASAN errors and leaks will follow...
test/boost/commitlog_test.cc(0): Leaving test case "test_commitlog_shutdown_during_wait"; testing time: 46614248us
test/boost/commitlog_test.cc(0): Test case "test_commitlog_deadlock_with_flush_threshold" is skipped because disabled
test/boost/commitlog_test.cc(0): Test case "test_commitlog_exceptions_in_allocate_ex" is skipped because disabled
test/boost/commitlog_test.cc(0): Test case "test_commitlog_exceptions_in_allocate_ex_deleted_file_no_recycle" is skipped because disabled
Leaving test module "test/boost/commitlog_test.cc"; testing time: 46614602us
ERROR 2022-06-15 07:45:53,225 [shard 0] commitlog - Could not delete segment : std::filesystem::__cxx11::filesystem_error (error system:2, filesystem error: remove failed: No such file or directory [��
)
Segmentation fault on shard 0.
Backtrace:
  0x3ad6c78
  0x3b083f6
  0x7f60e3f06a1f
  0x3a99d67
  0x23c9130
  0x3ae7bc4
  0x3ae8fa7
  0x3ae81fc
  0x3e65e08
  0x3e652e1
  0x3e6126d
  0x3acbafa
  /lib64/libpthread.so.0+0x92a4
  /lib64/libc.so.6+0x100322
=== TEST.PY SUMMARY START ===
Test exited with code -11
```

Superficially, this looks like use after free or use after move of the removed file name.

**Comment by bhalevy** [comment]
I guess this can be explained by
```
test/boost/commitlog_test.cc(880): fatal error: in "test_commitlog_shutdown_during_wait": log shutdown timed out. maybe it is deadlocked... Will not free log. ASAN errors and leaks will follow...
```

Why is that?

**Comment by elcallio** [comment]
Why the test deadlocks? No idea. Why the extra failure after? Because we probably killed the tempdir var before the commitlog var, and thus deleted the files it thought to do something with, what with it not being shut down and all. 
And unless this reproduces, I don't think there is all that much I can do with it...

**Comment by bhalevy** [comment]
Maybe the test should abort / assert instead of failing gracefully so we can try to figure out the deadlock from the coredump.

**Comment by bhalevy** [comment]
> we probably killed the tempdir var before the commitlog var, and thus deleted the files it thought to do something with

That would cause remove_file to fail indeed, but is there use-after-free of the filename.
I'd like to understand and fix that too.

**Comment by bhalevy** [comment]
Also the test code is:
```
    tmpdir tmp;
    cfg.commit_log_location = tmp.path().string();
    auto log = co_await commitlog::create_commitlog(cfg);
```

I don't see why the tmpdir would be destroyed before the commit log

**Comment by bhalevy** [comment]
As a general comment, (cc @xemul, @avikivity), our shutdown model is unclear and implemented differently by different services.
I suggest to define shutdown similar to abort (or deprecate the interface and just use a sharded abort source for shutdown), i.e. a single shot trigger event (that returns a future but only for signaling all shards).
The service stop() method should wait on the service to complete its shutdown process.
So in this case shutdown would return immediately, and if anything will hang, it would be stop(), not shutdown().

**Comment by bhalevy** [comment]
It looks like db::commitlog::segment_manager::shutdown continues to run in the background after the test case returns (and the commitlog is destroyed)

**Comment by elcallio** [comment]
Yes. It does. If the test deadlocked. Because it cannot finish. You are right about the tempdir, so then it is just a case of destroying a variable (commitlog) while it has pending operations. The test is for a fatal fail, it is not strange that the failing is fatal. Don't see a way to _undeadlock_ the deadlock and do a clean shutdown.

**Comment by bhalevy** [comment]
Longer term, I think that commitlog should have a stop method that would trigger shutdown if needed and wait on it (e.g. co_await _shutdown_promise->get_shared_future()), while shutdown should just trigger the shutdown process but not necessarily wait for it to stop.  Then in the test you can have a `auto stop_commitlog deferred_stop(log) call before destroying it to make sure it's always stopped before it's destroyed.  With that we can put the timeout on the deferred stop(), and abort if it expired.

**Comment by elcallio** [comment]
How would that help?

**Comment by bhalevy** [comment]
Happened here too: https://jenkins.scylladb.com/job/scylla-enterprise/job/next/2094/artifact/testlog/x86_64_release/commitlog_test.test_commitlog_shutdown_during_wait.1888.log

**Comment by bhalevy** [comment]
> Already done in refactor branch. Which is in PR limbo.

Let's not depend on it, those flaky failures are wasting time for meany people.
Can you please add an assertion or explicit abort when shutdown hang so we get a better chance at understanding and fixing it?

**Comment by kostja** [comment]
Apparently addressed by https://github.com/scylladb/scylladb/commit/aab7794c315d94df108f402b15586e87cc743c37 , please reopen if it fails again.
